### ISSUE TITLE
[FEAT][BATCH_NORM] Implement TileLang kernel and PyTorch-compatible Op for Batch Normalization

### ISSUE THREAD

**Issue by michaelwithu** [tile-ai/TileOPs#303]
## Background

Batch Normalization (BN) is a foundational technique in deep neural networks, introduced in [Ioffe & Szegedy (2015)](https://arxiv.org/pdf/1502.03167). It normalizes activations across the batch dimension to reduce internal covariate shift, enabling higher learning rates and more stable training.

PyTorch exposes BN through a unified backend via `torch.nn.functional.batch_norm`, which powers:
- `torch.nn.BatchNorm1d` — input shape `(N, C)` or `(N, C, L)`
- `torch.nn.BatchNorm2d` — input shape `(N, C, H, W)`
- `torch.nn.BatchNorm3d` — input shape `(N, C, D, H, W)`

All three normalize over the `(N, *)` dimensions for each channel `C`.

TileOPs currently has no Batch Normalization implementation. This issue proposes adding a high-performance BN kernel and Op following the existing 2-layer architecture.

---

## Algorithm Summary

**Forward pass (training mode):**

```
# For each channel c:
mu_c    = mean(x[:, c, ...])                     # reduce over (N, *)
sigma_c = sqrt(var(x[:, c, ...]) + eps)          # reduce over (N, *)
x_hat   = (x - mu_c) / sigma_c                  # elementwise normalize
y       = gamma * x_hat + beta                   # scale and shift

# Update running statistics:
running_mean = (1 - momentum) * running_mean + momentum * mu_c
running_var  = (1 - momentum) * running_var  + momentum * var_c
```

**Forward pass (inference mode):**

```
y = gamma * (x - running_mean) / sqrt(running_var + eps) + beta
```

**Backward pass:**

```
# Gradients w.r.t. learnable parameters:
d_gamma = sum(grad_out * x_hat, dim=(N, *))
d_beta  = sum(grad_out,         dim=(N, *))

# Gradient w.r.t. input (requires mean/var from forward):
dx_hat  = grad_out * gamma
d_var   = sum(dx_hat * (x - mu) * -0.5 * (var + eps)^{-3/2}, dim=(N, *))
d_mu    = sum(dx_hat * -1/sigma + d_var * mean(-2*(x-mu)), dim=(N, *))
d_x     = dx_hat / sigma + d_var * 2*(x - mu)/m + d_mu/m
```

---

## Proposed Implementation

Following the TileOPs 2-layer architecture:

### Layer 1 — TileLang Kernels (`tileops/kernels/batch_norm/`)

| File | Description |
|------|-------------|
| `batch_norm_fwd.py` | Forward kernel: online Welford reduction + normalize + write running stats |
| `batch_norm_bwd.py` | Backward kernel: two-pass reduction for `d_gamma`, `d_beta`, and `d_x` |

**Key design decisions:**
- Use **Welford's online algorithm** for numerically stable single-pass mean+variance computation
- **Vectorized loads** (e.g., 128-bit) along the spatial/feature dimension
- **Shared memory reduction** across warps within each CTA
- Separate **training** and **inference** code paths (inference path can fuse into a single elementwise pass)
- Support `float16` / `bfloat16` inputs with `float32` accumulation for numerical stability

### Layer 2 — PyTorch Op (`tileops/ops/batch_norm.py`)

```python
class BatchNormFwdOp(Op):
    """Stateless batch normalization forward operator."""

    def forward(self, x, weight, bias, running_mean, running_var,
                training, momentum, eps) -> Tensor:
        ...

class BatchNormBwdOp(Op):
    """Stateless batch normalization backward operator."""

    def forward(self, grad_out, x, weight, mean, rstd) -> Tuple[Tensor, Tensor, Tensor]:
        ...
```

### Tests and Benchmarks

- `tests/ops/test_batch_norm.py` — correctness vs `torch.nn.functional.batch_norm`
- `benchmarks/ops/bench_batch_norm.py` — latency and bandwidth vs PyTorch (cuDNN backend)

**Test shapes to cover:**

| Variant | Shape |
|---------|-------|
| BatchNorm1d | `(N=32, C=512)`, `(N=32, C=512, L=1024)` |
| BatchNorm2d | `(N=32, C=256, H=56, W=56)` (typical ResNet) |
| BatchNorm3d | `(N=8, C=64, D=16, H=32, W=32)` |

---

## Expected Impact

Batch Normalization is a compute-bound and memory-bandwidth-bound operation. A fused TileLang kernel avoids the multi-kernel dispatch overhead of cuDNN and enables:

- **Fusion opportunities**: future `Conv + BN + ReLU` or `Linear + BN` fused kernels
- **Custom quantization-aware BN**: e.g., BN with FP8 accumulation for modern inference pipelines
- **Foundation for LayerNorm / GroupNorm variants** following the same reduction pattern

---

## Next Steps

- [ ] Implement `batch_norm_fwd.py` kernel with Welford reduction (training mode first)
- [ ] Implement `batch_norm_bwd.py` kernel
- [ ] Add inference-mode fast path (single elementwise pass)
- [ ] Write `BatchNormFwdOp` and `BatchNormBwdOp` wrappers
- [ ] Add correctness tests against `torch.nn.functional.batch_norm`
- [ ] Add benchmarks comparing against cuDNN baseline

A PR will follow after this issue is confirmed.
